File: leetcode/hard/0099RecoverBST/golang/recover.go
package golang

// TreeNode Data Class
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func recoverTree(root *TreeNode) {
	var first, second, prev *TreeNode
	var dfs func(*TreeNode)
	dfs = func(root *TreeNode) {
		if root.Left != nil {
			dfs(root.Left)
		}

		if prev != nil && prev.Val > root.Val {
			if first == nil {
				first = prev
			}
			if first != nil {

				second = root
			}
		}

		prev = root

		if root.Right != nil {
			dfs(root.Right)
		}
	}

	dfs(root)
	first.Val, second.Val = second.Val, first.Val
}

File: leetcode/hard/0099RecoverBST/recover.cpp
#include <bits/stdc++.h>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    TreeNode* CurMax = new TreeNode(INT_MIN);
    TreeNode* first = NULL;
    TreeNode* second = NULL;

public:
    void recoverTree(TreeNode* root) 
    {
       DFS(root);
       
       int temp = first->val;
       first->val = second->val;
       second->val = temp;
    }
    
    void DFS(TreeNode* node)
    {
        if (node==NULL) return;
        
        DFS(node->left);
        
        if (node->val >= CurMax->val)
            CurMax = node;
        else
        {
            if (first==NULL) 
            {
                first=CurMax;
                second=node;
                CurMax=node;
            }
            else
                second=node;
        }
        
        DFS(node->right);
            
    }
    
};
